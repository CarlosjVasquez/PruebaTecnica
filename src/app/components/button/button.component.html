<button
  [type]="type"
  class="btn btn-{{ color }} btn-{{ size }}"
  [class.is-btn-disabled]="disabled"
  (click)="onHandleClick()"
  [disabled]="disabled"
>
  {{ label }}
</button>
