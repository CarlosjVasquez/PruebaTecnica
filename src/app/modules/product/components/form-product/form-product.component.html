<div class="mt-3">
  <form [formGroup]="form" (ngSubmit)="onSubmit()">
    <div class="row">
      <div class="col">
        <custom-input
          label="ID"
          id="id"
          name="id"
          [onlyLetters]="true"
          [disabled]="edit"
          [formControl]="formControlType('id')"
        />
      </div>
      <div class="col">
        <custom-input
          label="Nombre"
          id="name"
          name="name"
          [minlength]="5"
          [maxlength]="100"
          [required]="true"
          [onlyLetters]="true"
          [formControl]="formControlType('name')"
        />
      </div>
    </div>
    <div class="row mt-3">
      <div class="col">
        <custom-input
          label="Descripción"
          id="description"
          name="description"
          [minlength]="10"
          [maxlength]="200"
          [required]="true"
          [onlyLetters]="true"
          [formControl]="formControlType('description')"
        />
      </div>
      <div class="col">
        <custom-input
          label="Logo"
          id="logo"
          name="logo"
          [required]="true"
          [formControl]="formControlType('logo')"
        />
      </div>
    </div>
    <div class="row mt-3">
      <div class="col">
        <custom-input
          type="date"
          label="Fecha de Liberación"
          id="date_release"
          name="date_release"
          [required]="true"
          [min]="currentDate()"
          (onChangeEventEmitter)="calcdate_revision()"
          [formControl]="formControlType('date_release')"
        />
      </div>
      <div class="col">
        <custom-input
          label="Fecha de Revisión"
          id="date_revision"
          name="date_revision"
          type="date"
          [required]="true"
          [disabled]="true"
          [formControl]="formControlType('date_revision')"
        />
      </div>
    </div>
    <div class="row mt-5">
      @if(!edit) {
      <div class="col">
        <custom-button
          type="reset"
          class="content-btn-right"
          label="Reiniciar"
          asda
          color="body"
        />
      </div>
      }

      <div class="col">
        <custom-button
          type="submit"
          [class]="edit ? 'content-btn-center' : 'content-btn-left'"
          label="Enviar"
          [disabled]="form.invalid"
        />
      </div>
    </div>
  </form>
</div>
