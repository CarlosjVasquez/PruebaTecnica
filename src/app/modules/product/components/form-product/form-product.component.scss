.content-btn-left {
  display: flex;
  justify-content: start;
  align-items: center;
}
.content-btn-right {
  display: flex;
  justify-content: end;
  align-items: center;
}

.content-btn-center {
  display: flex;
  justify-content: center;
  align-items: center;
}
