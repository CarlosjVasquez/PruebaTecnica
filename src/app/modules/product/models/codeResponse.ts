export interface ResponseSuccessfully<T> {
  message: string;
  data: T | T[];
}

export interface BadRequestError {
  name: string;
  message: string;
}
