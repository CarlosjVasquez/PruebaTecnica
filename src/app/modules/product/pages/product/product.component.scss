@import "../../../../../styles/variables";

.title {
  border-bottom: 1px map-get($colors, "secondary") solid;
  text-align: center;
}

.content {
  position: relative;
}

.content-x {
  position: absolute;
  top: 1rem;
  right: 1rem;
}

.close {
  color: map-get($colors, "gray");
  float: right;
  font-size: 28px;
  font-weight: bold;
  cursor: pointer;
}

.close:hover,
.close:focus {
  color: map-get($colors, "dark");
  text-decoration: none;
  cursor: pointer;
}
