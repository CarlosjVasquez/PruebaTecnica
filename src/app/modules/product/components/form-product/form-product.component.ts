import { Component, EventEmitter, Input, OnInit, Output } from '@angular/core';
import { InputComponent } from '@components/input/input.component';
import {
  FormBuilder,
  FormControl,
  FormGroup,
  FormsModule,
  NgForm,
  ReactiveFormsModule,
  Validators,
} from '@angular/forms';
import { Product } from '../../models/product';
import { ButtonComponent } from '@components/button/button.component';
import { ProductService } from '../../services/product.service';
import { debounceTime, filter, Subscription, switchMap } from 'rxjs';

@Component({
  selector: 'form-product',
  standalone: true,
  imports: [InputComponent, FormsModule, ButtonComponent, ReactiveFormsModule],
  templateUrl: './form-product.component.html',
  styleUrl: './form-product.component.scss',
})
export class FormProductComponent implements OnInit {
  @Input() edit: boolean = false;
  @Output() submitEventEmitter: EventEmitter<Product> = new EventEmitter();
  private readonly subscriptionSelectedProduct: Subscription;
  private selectedProduct: Product = new Product('', '', '', '', '', '');
  form!: FormGroup;

  constructor(
    private readonly productService: ProductService,
    private readonly _fb: FormBuilder
  ) {
    this.subscriptionSelectedProduct =
      this.productService.selectedProduct$.subscribe((value) => {
        this.selectedProduct = value;
      });
  }

  ngOnInit(): void {
    this.initForm();
  }

  initForm() {
    this.form = this._fb.group({
      id: new FormControl(this.selectedProduct.id, [Validators.minLength(3), Validators.maxLength(10), Validators.required]),
      name: new FormControl(this.selectedProduct.name),
      description: new FormControl(this.selectedProduct.description),
      logo: new FormControl(this.selectedProduct.logo),
      date_release: new FormControl(this.selectedProduct.date_release),
      date_revision: new FormControl(this.selectedProduct.date_revision),
    });
    this.form.get('id')?.valueChanges.pipe(
      debounceTime(500),
      filter(id => id?.length > 2),
      switchMap(id => this.productService.verificationID(id))
    ).subscribe({
      next: (result) => {
        const control = this.form.get('id');
        if (result) {
          const currentErrors = control?.errors ? { ...control.errors } : {};
          currentErrors['idValidation'] = true;
          control?.setErrors(currentErrors);
        } else if (control?.errors) {
          const currentErrors = { ...control.errors };
          delete currentErrors['idValidation'];
          control?.setErrors(Object.keys(currentErrors).length ? currentErrors : null);
        } else {
          control?.setErrors(null);
        }

      }

    });
  }

  onSubmit() {
    this.submitEventEmitter.emit(this.form.value);
  }

  getSearchErrors(form: NgForm) {
    const searchControl = form.controls['search'];
    return searchControl ? searchControl.errors : null;
  }

  currentDate(): string {
    return new Date().toISOString().split('T')[0];
  }

  calcdate_revision(): void {
    const date_release = new Date(this.form.value.date_release);
    date_release.setFullYear(date_release.getFullYear() + 1);
    this.form.patchValue({
      date_revision: date_release.toISOString().split('T')[0],
    });
  }

  formControlType(key: string) {
    return this.form.controls[key] as FormControl;
  }

  ngOnDestroy(): void {
    this.subscriptionSelectedProduct.unsubscribe();
  }
}
