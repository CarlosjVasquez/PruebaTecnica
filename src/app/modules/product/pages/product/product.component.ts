import { Component, OnInit } from '@angular/core';
import { FormProductComponent } from '../../components/form-product/form-product.component';
import { Router } from '@angular/router';
import { Product } from '../../models/product';
import { ProductService } from '../../services/product.service';

@Component({
  selector: 'app-product',
  standalone: true,
  imports: [FormProductComponent],
  templateUrl: './product.component.html',
  styleUrl: './product.component.scss',
})
export class ProductComponent implements OnInit {
  title: string = 'Formulario de Registro';
  edit: boolean = false;

  constructor(private readonly router: Router, private readonly productService: ProductService) {
    if (this.router.url === '/product/update') this.edit = true;
  }

  ngOnInit(): void {
    if (this.edit) {
      this.title = 'Actualizar Producto';
    }
  }

  clearSelectProduct() {
    this.productService.selectProduct(new Product('', '', '', '', '', ''));
    this.router.navigate(['/products']);
  }

  onSubmitForm(product: Product) {
    this.edit
      ? this.productService.updateProduct(product)
      : this.productService.addProduct(product);
    this.clearSelectProduct();
  }
}
