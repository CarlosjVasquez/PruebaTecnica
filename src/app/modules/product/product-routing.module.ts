import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import { ListProductComponent } from './pages/list-products/list-product.component';
import { ProductComponent } from './pages/product/product.component';

const routes: Routes = [
  {
    path: '',
    component: ListProductComponent,
  },
  {
    path: 'create',
    component: ProductComponent,
  },
  {
    path: 'update',
    component: ProductComponent,
  },
];

@NgModule({
  imports: [RouterModule.forChild(routes)],
  exports: [RouterModule]
})
export class ProductRoutingModule { }
