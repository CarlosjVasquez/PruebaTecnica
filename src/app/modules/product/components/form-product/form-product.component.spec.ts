import { ComponentFixture, TestBed } from '@angular/core/testing';

import { FormProductComponent } from './form-product.component';
import { By } from '@angular/platform-browser';
import { of } from 'rxjs';
import { Product } from '../../models/product';
import { ProductService } from '../../services/product.service';
import { ButtonComponent } from '../../../../components/button/button.component';
import { InputComponent } from '../../../../components/input/input.component';
import { FormsModule, NgForm } from '@angular/forms';
import { provideHttpClient } from '@angular/common/http';
import { provideHttpClientTesting } from '@angular/common/http/testing';

describe('FormProductComponent', () => {
  let component: FormProductComponent;
  let fixture: ComponentFixture<FormProductComponent>;
  let compiled: HTMLElement;
  let productService: ProductService;

  interface MockFormControls {
    [key: string]: any;
  }

  beforeEach(async () => {
    const productServiceMock = {
      verificationID: jest.fn().mockReturnValue(of(false)),
    };

    await TestBed.configureTestingModule({
      providers: [
        FormsModule,
        FormProductComponent,
        InputComponent,
        ButtonComponent,
        provideHttpClient(),
        provideHttpClientTesting(),
      ],
    }).compileComponents();

    productService = TestBed.inject(ProductService);
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FormProductComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(FormProductComponent);
    component = fixture.componentInstance;
    component.product = new Product(
      '1',
      'Product 1',
      'Description',
      'logo.png',
      '2023-01-01',
      '2024-01-01'
    );
    fixture.detectChanges();
  });

  test('should create', () => {
    expect(component).toBeTruthy();
  });

  test('match to snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });

  test('should button', () => {
    const button = fixture.debugElement.query(By.directive(ButtonComponent));
    expect(button).toBeTruthy();
  });

  test('should input', () => {
    const input = fixture.debugElement.query(By.directive(InputComponent));
    expect(input).toBeTruthy();
  });

  test('should emit submit event with product data on submit', () => {
    jest.spyOn(component.submitEventEmitter, 'emit');
    component.onSubmit();
    expect(component.submitEventEmitter.emit).toHaveBeenCalledWith(
      component.product
    );
  });

  test('should return search control errors from the form', () => {
    const mockForm = {
      controls: {
        search: {
          errors: { required: true },
        },
      },
    } as unknown as NgForm & MockFormControls;
    const errors = component.getSearchErrors(mockForm as NgForm);
    expect(errors).toEqual({ required: true });
  });

  test('should return null if search control is not present in the form', () => {
    const mockForm = {
      controls: {},
    } as unknown as NgForm & MockFormControls;
    const errors = component.getSearchErrors(mockForm as NgForm);
    expect(errors).toBeNull();
  });

  test('should return the current date in YYYY-MM-DD format', () => {
    const currentDate = new Date().toISOString().split('T')[0];
    expect(component.currentDate()).toEqual(currentDate);
  });

  test('should calculate date_revision as one year after date_release', () => {
    component.product.date_release = '2023-01-01';
    component.calcdate_revision();
    expect(component.product.date_revision).toBe('2024-01-01');
  });

  test('should set errorVerificationId if verificationID returns true', () => {
    jest.spyOn(productService, 'verificationID').mockReturnValue(of(true));
    component.verificationID('test-id');
    fixture.detectChanges();
    expect(component.errorVerificationId).toEqual({
      customError: { message: 'ID no vÃ¡lido!' },
    });
  });

  test('should set errorVerificationId to null if verificationID returns false', () => {
    jest.spyOn(productService, 'verificationID').mockReturnValue(of(false));
    component.verificationID('test-id');
    fixture.detectChanges();
    expect(component.errorVerificationId).toBeNull();
  });

  test('should call verificationID method of ProductService', () => {
    const productId = 'test-id';
    jest.spyOn(productService, 'verificationID').mockReturnValue(of(false));

    component.verificationID(productId);

    expect(productService.verificationID).toHaveBeenCalledWith(productId);
  });
});
