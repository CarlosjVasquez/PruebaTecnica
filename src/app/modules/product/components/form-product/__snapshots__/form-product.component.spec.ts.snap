// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormProductComponent match to snapshot 1`] = `
<div
  id="root2"
>
  <div
    class="mt-3"
  >
    <form
      class="ng-untouched ng-pristine ng-valid"
      novalidate=""
    >
      <div
        class="row"
      >
        <div
          class="col"
        >
          <custom-input
            class="ng-untouched ng-pristine ng-valid"
            id="id"
            label="ID"
            maxlength="10"
            minlength="3"
            name="id"
            required=""
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="id"
              >
                ID
              </label>
              
              <input
                class="form-input ng-untouched ng-pristine ng-invalid"
                id="id"
                max="undefined"
                maxlength="10"
                min="undefined"
                minlength="3"
                placeholder=""
                required=""
                type="text"
              />
              
            </div>
          </custom-input>
        </div>
        <div
          class="col"
        >
          <custom-input
            class="ng-untouched ng-pristine ng-valid"
            id="name"
            label="Nombre"
            maxlength="100"
            minlength="5"
            name="name"
            required=""
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="name"
              >
                Nombre
              </label>
              
              <input
                class="form-input ng-untouched ng-pristine ng-invalid"
                id="name"
                max="undefined"
                maxlength="100"
                min="undefined"
                minlength="5"
                placeholder=""
                required=""
                type="text"
              />
              
            </div>
          </custom-input>
        </div>
      </div>
      <div
        class="row mt-3"
      >
        <div
          class="col"
        >
          <custom-input
            class="ng-untouched ng-pristine ng-valid"
            id="description"
            label="Descripción"
            maxlength="200"
            minlength="10"
            name="description"
            required=""
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="description"
              >
                Descripción
              </label>
              
              <input
                class="form-input ng-untouched ng-pristine ng-invalid"
                id="description"
                max="undefined"
                maxlength="200"
                min="undefined"
                minlength="10"
                placeholder=""
                required=""
                type="text"
              />
              
            </div>
          </custom-input>
        </div>
        <div
          class="col"
        >
          <custom-input
            class="ng-untouched ng-pristine ng-valid"
            id="logo"
            label="Logo"
            name="logo"
            required=""
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="logo"
              >
                Logo
              </label>
              
              <input
                class="form-input ng-untouched ng-pristine ng-invalid"
                id="logo"
                max="undefined"
                min="undefined"
                placeholder=""
                required=""
                type="text"
              />
              
            </div>
          </custom-input>
        </div>
      </div>
      <div
        class="row mt-3"
      >
        <div
          class="col"
        >
          <custom-input
            class="ng-untouched ng-pristine ng-valid"
            id="date_release"
            label="Fecha de Liberación"
            name="date_release"
            required=""
            type="date"
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="date_release"
              >
                Fecha de Liberación
              </label>
              
              <input
                class="form-input ng-untouched ng-pristine ng-invalid"
                id="date_release"
                max="undefined"
                min="2024-07-11"
                placeholder=""
                required=""
                type="date"
              />
              
            </div>
          </custom-input>
        </div>
        <div
          class="col"
        >
          <custom-input
            class="ng-untouched ng-pristine ng-valid"
            id="date_revision"
            label="Fecha de Revisión"
            name="date_revision"
            required=""
            type="date"
          >
            <div
              class="form-group"
            >
              <label
                class="form-label is-disabled"
                for="date_revision"
              >
                Fecha de Revisión
              </label>
              
              <input
                class="form-input ng-untouched ng-pristine ng-invalid"
                disabled=""
                id="date_revision"
                max="undefined"
                min="undefined"
                placeholder=""
                required=""
                type="date"
              />
              
            </div>
          </custom-input>
        </div>
      </div>
      <div
        class="row mt-5"
      >
        <div
          class="col"
        >
          <custom-button
            asda=""
            class="content-btn-right"
            color="body"
            label="Reiniciar"
            type="reset"
          >
            <button
              class="btn btn-body btn-md"
              type="reset"
            >
               Reiniciar

            </button>
          </custom-button>
        </div>
        
        <div
          class="col"
        >
          <custom-button
            class="content-btn-left"
            label="Enviar"
            type="submit"
          >
            <button
              class="btn btn-md btn-primary"
              type="submit"
            >
               Enviar

            </button>
          </custom-button>
        </div>
      </div>
    </form>
  </div>
</div>
`;
